[red-knot] Diagnostics for incorrect `bool` usages

MichaReiser opened the pull request:
## Summary

The diagnostic handling isn't ideal yet but I'd like to have some first feedback on the overall aproach (and maybe you've suggestions on how to improve the diagnostics). 


This PR introduces new `Type::try_iterate` and `Type::try_bool` methods that return a `Result` instead of always assuming that the operation was successful. ~~This PR also renames the existing `call` and `dunder_call` methods to `try_call` and `try_dunder_call` to match this pattern~~ (see https://github.com/astral-sh/ruff/pull/16261)


This PR also introudces the new lint rule `unsupported-bool-conversion` that detects implicit `bool` conversions where the type doesn't implement `__bool__` correctly.

## Performance

~~Codspeed shows a 4% perf regression on the cold benchmark. I suspect that this is due to Red Knot calling `bool` much more frequently than it used to.~~

This is no longer the case, thanks to @sharkdp's performance improvement to `try_bool`

## Test Plan

There's more to do but there are some new tests.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3083,12 +3117,100 @@ impl<'db> IterationOutcome<'db> {
+    /// The type has a `__bool__` attribute, but it doesn't have the right signature.
+    IncorrectSignature {
+        not_boolable_type: Type<'db>,
+        return_type: Type<'db>,
```
Creating a multi-span diagnostic that shows the return type probably requires us to return an entire `binding` (or `CallOutcome`) here.

MichaReiser commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -12,7 +12,7 @@ pub(super) use bind::{bind_call, CallBinding};
 /// It's guaranteed that the wrapped bindings have no errors.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub(super) enum CallOutcome<'db> {
+pub(crate) enum CallOutcome<'db> {
```
I should revert this

MichaReiser commented:
Huh, the perf regression is rather surprising.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -1436,76 +1444,109 @@ impl<'db> Type<'db> {
+                    let first_element = union_elements.next().unwrap();
+                    let mut truthiness = first_element.bool(db)?;
+
+                    for element in union_elements {
+                        if element.bool(db)? != truthiness {
+                            truthiness = Truthiness::Ambiguous;
+                        }
+                    }
```
I suspect that the regression comes from the fact that we no longer short-circuit if any union element inferred `Ambigious`.

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/call/builtins.md:
```diff
@@ -0,0 +1,14 @@
+class NotBool:
+    __bool__ = None
+
+# TODO: We should emit an `invalid-argument` error here for `2` because `bool` only takes one argument.
```
this is more something I noticed and considered worth adding a test. It isn't itself related to making `Type::bool` faliable

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/snapshots/membership_test.md_-_Comparison:_Membership_Test_-_Not-boolable_return_type.snap:
```diff
@@ -0,0 +1,53 @@
+ 9 | 10 in WithContains() and True
+10 | # error: [not-boolable] "Object of type `NotBoolable` can not be converted to a bool." 
+11 | 10 not in WithContains() or False 
+   | ^^^^^^^^^^^^^^^^^^^^^^^^ Object of type `NotBoolable` can not be converted to a bool.
```
The error messages here (and in other places where `bool` is called implicitly could be better but I'm not sure how they should look exactly. What makes them not very useful right now is that it's unclear where the `NotBoolable` comes from.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -1432,80 +1432,140 @@ impl<'db> Type<'db> {
+        &self,
+        db: &'db dyn Db,
+        allow_short_circuit: bool,
+    ) -> Result<Truthiness, BoolError<'db>> {
+        let truthiness =
```
The main change here is that this now returns a `Result`. The other change is in how `Type::Instnace` is handled.

AlexWaygood commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3035,47 +3116,126 @@ pub enum TypeVarBoundOrConstraints<'db> {
+}
+
+impl BoolError<'_> {
+    #[allow(clippy::unused_self)]
+    pub(super) fn fallback_truthiness(&self) -> Truthiness {
+        Truthiness::Ambiguous
+    }
```
hrm, I think I'd prefer to add a `Truthiness::from_bool_error()` function and use that

AlexWaygood replied:
Yeah, I see the consistency argument... I still feel like I'd prefer something inconsistent here than having a method that "pretends to be an instance method" but doesn't actually use `self`. That said, I don't feel very strongly.

Could we at least add a comment about why we're deliberately leaving `self` unused here, if we leave it as-is, though?

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -2375,7 +2393,12 @@ impl<'db> TypeInferenceBuilder<'db> {
-        self.infer_standalone_expression(test);
+        let test_ty = self.infer_standalone_expression(test);
+
+        if let Err(err) = test_ty.try_bool(self.db()) {
+            err.report_diagnostic(&self.context, &**test);
+        }
```
this is a very nice API :-)

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3363,8 +3414,6 @@ impl<'db> TypeInferenceBuilder<'db> {
                             bindings: _,
                             errors,
                         } => {
-                            // TODO: Remove the `Vec::from` call once we use the Rust 2024 edition
-                            //  which adds `Box<[T]>::into_iter`
```
...did they remove it again? :P

MichaReiser replied:
Why am I so terrible lately with rebasing/stashing my changes. No, I figured out how to avoid the `Vec::from` but the change got lost. Thanks for catching this.

AlexWaygood commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4507,11 +4627,14 @@ impl<'db> TypeInferenceBuilder<'db> {
                                 Type::Tuple(lhs),
                                 ast::CmpOp::Eq,
-                                *ty,
+                                *ty, range
                             ).expect("infer_binary_type_comparison should never return None for `CmpOp::Eq`");
```
```suggestion
                                *ty,
                                range,
                            ).expect("infer_binary_type_comparison should never return None for `CmpOp::Eq`");
```

MichaReiser replied:
Hmm thanks. `infer` and `types.rs` are a bit much for Rustfmt.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/instances/membership_test.md:
```diff
@@ -160,3 +160,24 @@ reveal_type(42 in A())  # revealed: bool
+    def __contains__(self, item) -> NotBoolable:
+        return NotBoolable()
+
+# error: [not-boolable] "Object of type `NotBoolable` can not be converted to a bool."
+10 in WithContains() and True
```
I think ideally the code and error message here would be something like

```suggestion
# error: [operator] "Cannot use `in` operator on object of type `WithContains` as its `__contains__` method returns `NotBoolable`, which cannot be evaluated in a boolean context"
10 in WithContains() and True
```

AlexWaygood replied:
> Slightly too long for my taste

For me too... but once we have the ability to add notes to diagnostics, I'd imagine it could become something like

```
error: [operator] cannot use `in` operator on object of type `WithContains`
    note: This is because `WithContains.__contains__` returns an instance of `NotBoolable`, which cannot be evaluated in a boolean context
    note: `NotBoolable` cannot be evaluated in a boolean context because its `__bool__` attribute is not callable
```

AlexWaygood replied:
See my comment at https://github.com/astral-sh/ruff/pull/16238#discussion_r1963014361. I think it's pretty important that we foreground *what's wrong with the specific code we're highlighting* in our diagnostic messages, and *what the user needs to do to resolve the error*. The invalid method definition is the underlying cause here, but the specific thing that's incorrect in the code we're highlighting with the diagnostic range is that the object with the invalid method is being used in a boolean context, and the specific way for the user to resolve that error is for them to stop doing that. Fixing the invalid method definition may be out of their power if it comes from a third-party library.

AlexWaygood replied:
To be clear: I don't think my disagreement with Carl should block this PR being merged. We can iterate on the error message and/or continue discussing this in an issue or on Discord.

MichaReiser replied:
I agree that this would be much better. 

I'm leaning towards not implementing this now. A) This feels like a task on its own because it isn't sufficient to prepend a message to every diagnostic; making it sound natural requires repeating much of the diagnostic code, and b) ideally, we'd have support from the diagnostic system to do this correctly.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/instances/rich_comparison.md:
```diff
@@ -315,3 +315,27 @@ def f(x: bool, y: int):
+
+<!-- snapshot-diagnostics -->
+
+Python implicitly calls `bool` on the comparison result of preceeding elements (but not for the last
```
```suggestion
Python implicitly calls `bool` on the comparison result of preceding elements (but not for the last
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/instances/rich_comparison.md:
```diff
@@ -315,3 +315,27 @@ def f(x: bool, y: int):
+# error: [not-boolable]
+10 < Comparable() < 20
+# error: [not-boolable]
+10 < Comparable() < Comparable()
```
Could we also test that we _don't_ emit a diagnostic if there are only two elements?

```suggestion
10 < Comparable() < Comparable()

Comparable() < Comparable()  # fine
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:
```diff
@@ -325,3 +325,25 @@ reveal_type(a is not c)  # revealed: Literal[True]
+b = (1, Comparable())
+
+# error: [not-boolable]
+a < b < b
```
```suggestion
a < b < b

a < b  # fine
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/expression/boolean.md:
```diff
@@ -101,3 +101,55 @@ reveal_type(bool([]))  # revealed: bool
+class NotBoolable:
+    __bool__: int
+
+# error: [not-boolable] "Object of type `NotBoolable` can not be converted to a bool because its `__bool__` attribute isn't callable."
```
really nice error message :-)

MichaReiser replied:
Yes, the union with `Unknown` messes everything up.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/narrow/truthiness.md:
```diff
@@ -107,6 +107,7 @@ more accurately.
 def flag() -> bool:
     return True
 
+# TODO: Not sure if this is correct but we're trying to convert `int` to a bool here? Is this intentional.
```
not sure I understand the TODO comment here?

MichaReiser replied:
Oh, this is outdated. That was a comment from before you helped me fix the `bool` implementation.

AlexWaygood reviewed (approved):
I have a bunch of nits but the overall approach looks great to me

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3035,47 +3116,126 @@ pub enum TypeVarBoundOrConstraints<'db> {
+                    &NOT_BOOLABLE,
+                    condition,
+                    format_args!(
+                        "Object of type `{}` can not be converted to a bool.",
```
```suggestion
                        "Object of type `{}` has an invalid `__bool__` and cannot be converted to a bool.",
```

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -2328,7 +2343,9 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         self.infer_target(target, iter, |db, iter_ty| {
-            iter_ty.iterate(db).unwrap_without_diagnostic()
+            // It's okay to ignore the `iterate` outcome here because
+            // `infer_for_statement_definition` reports a diagnostic if `iter_ty` isn't iterable.
```
I think this isn't quite right, and there should be a TODO here (though a low-priority one). If the `for` statement is not a definition, because the target is not a `Name` (Python allows e.g. `for x.attr in ...` and `for x[i] in ...`, strange as they look), we will never call `infer_for_statement_definition`, and we should ideally check the iterable for iterability and emit a diagnostic if it's not iterable.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3141,9 +3169,10 @@ impl<'db> TypeInferenceBuilder<'db> {
-            iterable_ty
-                .iterate(self.db())
-                .unwrap_with_diagnostic(&self.context, iterable.into())
+            iterable_ty.try_iterate(self.db()).unwrap_or_else(|err| {
+                err.report_diagnostic(&self.context, iterable.into());
+                err.fallback_element_type()
```
This is a common pattern and it would be nice if it could be a bit less verbose, but I guess there's no obvious way to do that short of a custom `Result` like type.

MichaReiser replied:
I had `fallback_element_type_with_diagnostic` but it just felt sooo long. I don't mind the repetitiveness here. It's simple and more flexible, e.g. you don't have to call `fallback_element_type` if you only want to report a diagnostic. 

Overall, I don't think we should spend much time on this: While repetitive, the call sites are very limited.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4082,27 +4142,35 @@ impl<'db> TypeInferenceBuilder<'db> {
                 );
 
-                match (ty.bool(db), is_last, op) {
+                let truthiness = ty.try_bool(self.db()).unwrap_or_else(|err| {
+                    if !is_last {
+                        err.report_diagnostic(&self.context, range);
+                    }
+
+                    err.fallback_truthiness()
+                });
+
+                if done {
+                    return Type::Never;
+                }
+
+                match (truthiness, is_last, op) {
```
I think this logic and the match below could be adjusted so we can avoid the unnecessary `try_bool` call on the last comparison entirely, instead of just skipping the diagnostic from it. Skipping the diagnostic is a tell that we are calling `try_bool` when we shouldn't need to because runtime doesn't.

If `is_last`, we always return `ty` and we don't actually care about `truthiness`.

I haven't checked the surrounding logic to see whether we enter this path even on single (non chained) comparisons, but if we do, fixing this could result in significantly fewer useless `try_bool` calls in real codebases.

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -4718,12 +4851,14 @@ impl<'db> TypeInferenceBuilder<'db> {
             match pairwise_eq_result {
                 // If propagation is required, return the result as is
                 todo @ Type::Dynamic(DynamicType::Todo(_)) => return Ok(todo),
+                // Using `bool` here is fine because we only use it to get to a truthiness
+                // but it doesn't mimic a runtime `bool` call.
```
This comment is not right, we are mimicking a runtime `bool` call here:

```
>>> class A:
...     def __eq__(self, other):
...         return B()
...
>>> class B:
...     def __bool__(self):
...         raise RuntimeError("boom")
...
>>> (A(),) == (A(),)
Traceback (most recent call last):
  File "<python-input-16>", line 1, in <module>
    (A(),) == (A(),)
  File "<python-input-13>", line 3, in __bool__
    raise RuntimeError("boom")
RuntimeError: boom
```

Because `__eq__` is allowed to return any object, Python at runtime has to check that object for truthiness, which could fail; just like we have to do here.

(Not saying we have to fix this in this PR, just that the comment here should be a TODO instead of an incorrect rationale.)

MichaReiser replied:
Nice catch. I tried to come up with a code example where we have to emit a diagnostic here but wasn't able to. Let me revisit this.

MichaReiser replied:
I think the code here is actually okay. The problem is in `rich_comparison` which doesn't handle possibly unboundness or `__bool__` correctly. I'll add a test and a todo.

https://github.com/astral-sh/ruff/blob/5faf0b9c2e7c536d4a3d272a646bb77b72e41ce4/crates/red_knot_python_semantic/src/types/infer.rs#L4781-L4782

carljm reviewed (approved):
This is great!! I left a lot of comments, but many of them are wording bikeshedding :)

AlexWaygood commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3035,47 +3116,126 @@ pub enum TypeVarBoundOrConstraints<'db> {
+                        &NOT_BOOLABLE,
+                        condition,
+                        format_args!(
+                            "Object of type `{}` can not be converted to a bool",
```
> I would prefer "has an invalid `__bool__` method" over either of "cannot be converted to a bool" or "cannot be evaluated in a boolean context." It is less vague and points more directly at the real problem.

But that wording implies that the only way for the user to resolve the diagnostic is to go and fix the invalid `__bool__` definition. It's not possible for the user to do that if it comes from some third-party library like numpy or pandas. I would say that in the majority of cases, the fix for the diagnostic is to stop using objects of that type in a boolean context rather than to go and fix the invalid `__bool__` method — and that's partly why we flag the usages of the method rather than the definition of the method with our diagnostics

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -1461,38 +1501,72 @@ impl<'db> Type<'db> {
-                    return Truthiness::Ambiguous;
+                let mut union_elements = union.elements(db).iter();
+                let first_element = union_elements.next().unwrap();
+                let mut truthiness = first_element.try_bool_impl(db, allow_short_circuit)?;
```
I'd argue it isn't incorrect, but the error is less helpful. `allow_short_circuit` is to return early on an ambiguous result (maybe I should rename the variable) but that could mean that the method returns `Ok` for a union even if one variant can't be booled.

MichaReiser commented on crates/red_knot_python_semantic/src/types/diagnostic.rs:
````diff
@@ -490,6 +491,29 @@ declare_lint! {
+    /// if NotBoolable():
+    ///     pass
+    /// ```
+    pub(crate) static NOT_BOOLABLE = {
````
I don't like the `invalid-bool-method` too much because I'd expect this diagnostic to be raised on the definition side and not where `__bool__` is used. 

For example, the following suppression comments feels out of place:

```
if NotBoolable(): # red-knot: ignore[invalid-bool-method]
```

I agree that there's a concern with `Boolable` not being a concept in typeshed.

MichaReiser replied:
I think this also comes back to our discussion here https://github.com/astral-sh/ruff/pull/16238#discussion_r1963014361 and I agree with @AlexWaygood that the most common fix is to not use this object in a bool context. Tricky

MichaReiser commented on crates/red_knot_python_semantic/src/types/diagnostic.rs:
```diff
@@ -490,6 +491,29 @@ declare_lint! {
+    /// class NotBoolable:
+    ///     __bool__ = None
+    ///
+    /// if NotBoolable():
+    ///     pass
```
Thanks. I didn't use the unpacking syntax to keep the example simple.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3035,47 +3116,126 @@ pub enum TypeVarBoundOrConstraints<'db> {
-    },
-    PossiblyUnboundDunderIter {
+enum IterateError<'db> {
+    /// The type isn't iterable because it doesn't implement the `__iter__` or `__get__item` "protocol".
```
Nice!

AlexWaygood commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3006,47 +3088,133 @@ pub enum TypeVarBoundOrConstraints<'db> {
+enum IterateError<'db> {
+    /// The type isn't iterable because it doesn't implement the new-style or old-style iteration protocol
+    ///
+    /// The new-style iteration protocol requies a type being iterated overto have an `__iter__`
```
```suggestion
    /// The new-style iteration protocol requires a type being iterated over to have an `__iter__`
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -3006,47 +3088,133 @@ pub enum TypeVarBoundOrConstraints<'db> {
+    ///
+    /// The new-style iteration protocol requies a type being iterated overto have an `__iter__`
+    /// method that returns something with a `__next__` method. The old-style iteration
+    /// protocol requies a type being iterated over to have a `__getitem__` method that accepts
```
```suggestion
    /// protocol requires a type being iterated over to have a `__getitem__` method that accepts
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/call/builtins.md:
```diff
@@ -0,0 +1,14 @@
+# TODO: We should emit an `invalid-argument` error here for `2` because `bool` only takes one argument.
+bool(1, 2)
+
+# TODO: We should emit a `not-boolable` error here because the argument doesn't implement `__bool__` correctly.
```
```suggestion
# TODO: We should emit a `unsupported-bool-conversion` error here because the argument doesn't implement `__bool__` correctly.
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/instances/membership_test.md:
```diff
@@ -160,3 +160,24 @@ reveal_type(42 in A())  # revealed: bool
+    def __contains__(self, item) -> NotBoolable:
+        return NotBoolable()
+
+# error: [unsupported-bool-conversion]
+10 in WithContains() and True
+# error: [unsupported-bool-conversion]
+10 not in WithContains() or False
```
(I still think that `operator` might actually be a better error code here specifically -- the invalid `__bool__` definition here means that you can't use the `in` operator with the `WithContains` type)

MichaReiser replied:
Closing as duplicate

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/instances/rich_comparison.md:
I agree with Carl's suggested changes in this file

MichaReiser commented:
Putting this back to draft or Alex keeps reviewing faster than I can address the comments ;)

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -1461,38 +1501,72 @@ impl<'db> Type<'db> {
+                            }
+                            Err(
+                                CallError::PossiblyUnboundDunderCall { .. }
+                                | CallError::Union { .. },
```
I added some specific handling for unions where exactly one variant fails.

AlexWaygood commented on crates/red_knot_test/src/parser.rs:
```diff
@@ -618,7 +622,7 @@ impl<'s> Parser<'s> {
             }
             None => match lang {
-                "py" => EmbeddedFilePath::Autogenerated(PySourceType::Python),
+                "py" | "python" => EmbeddedFilePath::Autogenerated(PySourceType::Python),
```
Not that it really matters much, but one reason why it is quite nice to stick with only using ```` ```py ```` or ```` ```pyi ```` is that currently all the tags we use exactly map to the file extension that the file will be saved as in the `MemorFileSystem`. If you use ```` ```py ````, it saves it with the `.py` file extension; if you use ```` ```pyi ````, it saves it with the `.pyi` file extension; if you use ```` ```toml ````, it saves it as a `.toml` file... but ```` ```python ```` here isn't going to cause mdtest to save the snippet with a `.python` extension.

MichaReiser replied:
Stop looking at my PR, it's in draft ;)

I was told yesterday that this wouldn't be hard to change but I precisely haven't changed it before because I was afraid of the push back and discussions it involves. 

I can revert the change here and open a separate PR but I'm very used to writing `python` and `RustRover` suggestions `python` but never `py`, making it very annoying because tests always fail.


I'd find this a quality-of-life improvement. I agree, it leads to less consistency but it's common that markdown editors/renderers support both: names and extension.

MichaReiser commented:
Okay, I think I addressed all feedback with the exception of special casing `in`. I decided to defer it because I suspect it's a fair bit of work on its own and the right approach also heavily depends on the functionality of our diagnostic system. That's why I'd rather wait. 

I did improve the error messages for unions.

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:
````diff
@@ -334,3 +334,43 @@ reveal_type(a is not c)  # revealed: Literal[True]
+a < b  # fine
+```
+
+## Equality with elements that incorrectly implement `__bool__`
````
```suggestion
## Tuple equality with elements that incorrectly implement `__bool__`

Python does not generally attempt to coerce the result of `==` and `!=` operations between two
arbitrary objects to a `bool`, but a comparison of tuples will fail if the result of comparing any
pair of elements at equivalent positions cannot be converted to a `bool`:
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/conditional/if_expression.md:
```diff
@@ -35,3 +35,13 @@ def _(flag: bool):
+class NotBoolable:
+    __bool__ = 3
+
+# error: [unsupported-bool-conversion] "Boolean conversion is unsupported for type `NotBoolable`; it's `__bool__` method isn't callable"
```
this should be "its" rather than "it's" because it's not a shortening of "it is"

```suggestion
# error: [unsupported-bool-conversion] "Boolean conversion is unsupported for type `NotBoolable`; its `__bool__` method isn't callable"
```

MichaReiser replied:
uh, this will be annoying

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:
```diff
@@ -334,3 +334,43 @@ reveal_type(a is not c)  # revealed: Literal[True]
 
 // TODO
+
+## Comparison with elements that incorrectly implement `__bool_`
```
```suggestion
## Chained comparisons with elements that incorrectly implement `__bool_`
```

AlexWaygood commented on crates/red_knot_python_semantic/resources/mdtest/comparison/tuples.md:
```diff
@@ -334,3 +334,43 @@ reveal_type(a is not c)  # revealed: Literal[True]
+## Comparison with elements that incorrectly implement `__bool_`
+
+<!-- snapshot-diagnostics -->
+
```
````suggestion

For an operation `A() < A()` to succeed at runtime, the `A.__lt__` method does not necessarily need
to return an object that is convertible to a `bool`. However, the return type _does_ need to be
convertible to a  `bool` for the operation `A() < A() < A()` (a *chained* comparison) to succeed.
This is because `A() < A() < A()` desugars to something like this, which involves several implicit
conversions to `bool`:

```ignore
def compute_chained_comparison():
		a1 = A()
		a2 = A()
		first_comparison = a1 < a2
		return first_comparison and (a2 < A())
```

````

MichaReiser replied:
And I can see, you're a fan of my `ignore` file type hehe

AlexWaygood replied:
> And I can see, you're a fan of my `ignore` file type hehe

haha yes, it's a great idea!

MichaReiser commented:
I'll merge this before we get this to over 100 comments ;) 97 is enough
